### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by saiik** [cloudcreativity/laravel-json-api#104]
Hello,

is there a way to optimize the performance? In our API we have a ~600ms response time for a resource with 12 relationships. 

We would like to archive something about 200ms response time. I've already refactored a lot of the adapters and stuff and got it already to 600ms from 1s.

Are there any tipps to get the best out of this package?

PS:

A resource with no relationship has already a response time of ~400ms.

**Comment by lindyhopchris** [comment]
Hi,

Great question. As the package is still in development, I haven't looked extensively into performance though this is something that should be done.

For a resource with no relationships, we're seeing ~300ms response times. That's acceptable for our use case though would be ideal if it could be reduced.

That time is with routes and config cached. Without either of those cached the response time will rise.

If you have a complex resource with lots of attributes, you'll probably get a performance bump if you write a Schema from scratch rather than using the Eloquent schema. Extend `CloudCreativity\LaravelJsonApi\Schema\AbstractSchema` instead. I'm not sure how much time that will save, but the Eloquent schema is processing things to work out what to serialize, whereas if you write one that is just explicit about what to serialize then that might save some time.

I'd be interested to hear your thoughts on where improvements could be gained, particularly if you've learnt anything by reducing times from 1s to 600ms. What did you change? Implementing any performance improvements would obviously be a priority thing to get into the package.

**Comment by saiik** [comment]
Hey,

thank you for your suggestion. I've tried to write a simple schema where only the model attributes will be displayed. It's at ~550ms now and without any virtual attributes it's at ~450ms. (This specific resource has 3 virtual attributes)

To the performance boost from 1s to 600ms:

I've made a base class for every adatper, hydrator, validator and model and this gained some performance. 

But i think we lose a lot of performance with our background stuff (verfiy jwt tokens, jwt scopes protected routes and attributes etc. etc.)
